test/framework/ownerreference_helpers.go

```diff
@@ -0,0 +1,260 @@
+	},
+	machineKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{machineSetGVK}, []schema.GroupVersionKind{kubeadmControlPlaneGVK})
+	},
+	machineHealthCheckKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+}
+
+// Kind and GVK for types in the exp package.
+var (
+	clusterResourceSetKind        = "ClusterResourceSet"
+	clusterResourceSetBindingKind = "ClusterResourceSetBinding"
+
+	clusterResourceSetGVK = addonsv1.GroupVersion.WithKind(clusterResourceSetKind)
+)
+var ExpOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	clusterResourceSetKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	clusterResourceSetBindingKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK, clusterResourceSetGVK})
+	},
+}
+
+var (
+	configMapKind = "ConfigMap"
+	secretKind    = "Secret"
+)
+
+var SecretOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{}
+var ConfigMapReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	configMapKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterResourceSetGVK}, []schema.GroupVersionKind{})
+	},
+	secretKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{kubeadmControlPlaneGVK}, []schema.GroupVersionKind{kubeadmConfigGVK})
+	},
+}
+
+// Kind and GVK for types in the Kubeadm ControlPlane package.
+var (
+	kubeadmControlPlaneKind         = "KubeadmControlPlane"
+	kubeadmControlPlaneTemplateKind = "KubeadmControlPlaneTemplate"
+
+	kubeadmControlPlaneGVK = controlplanev1.GroupVersion.WithKind(kubeadmControlPlaneKind)
+)
+var KubeadmControlPlaneOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	kubeadmControlPlaneKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+	kubeadmControlPlaneTemplateKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterClassGVK})
+	},
+}
+
+// Kind and GVK for types in the Kubeadm Bootstrap package.
+var (
+	kubeadmConfigKind         = "KubeadmConfig"
+	kubeadmConfigTemplateKind = "KubeadmConfigTemplate"
+
+	kubeadmConfigGVK = bootstrapv1.GroupVersion.WithKind(kubeadmConfigKind)
+)
+
+var KubeadmBootstrapOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	kubeadmConfigTemplateKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK}, []schema.GroupVersionKind{clusterClassGVK})
```

Comment:
I think I understand why we have two options here (kubeadmConfigTemplate in a ClusterClass or in Cluster), but I'm wondering if there is a way that we can use to make more specific assertions, might be using the concept of tenant that exists in clusterctl discovery...
Same below for the dockerMachineTemplate or for everything were we use hasOneOfExactOwnersByGVK